framework/src/timesteppers/Stepper.C

```diff
@@ -0,0 +1,748 @@
+/*                                                              */
+/*            See COPYRIGHT for full restrictions               */
+/****************************************************************/
+
+#include "Stepper.h"
+
+StepperInfo::StepperInfo()
+  : _step_count(1),
+    _time(0),
+    _dt(),
+    _nonlin_iters(0),
+    _lin_iters(0),
+    _converged(),
+    _solve_time_secs(),
+    _soln_nonlin(nullptr),
+    _soln_aux(nullptr),
+    _soln_predicted(nullptr),
+    _snapshot(false),
+    _rewind(false),
+    _rewind_time(-1),
+    _dummy_comm()
+{
+  for (int i = 0; i < 3; i++)
+  {
+    _dt.push_front(0);
+    _converged.push_front(true);
+    _solve_time_secs.push_front(0);
+  }
+  _soln_nonlin.reset(NumericVector<Number>::build(_dummy_comm).release());
+  _soln_nonlin->init(0, 0, false, SERIAL);
+  _soln_aux.reset(NumericVector<Number>::build(_dummy_comm).release());
+  _soln_aux->init(0, 0, false, SERIAL);
+  _soln_predicted.reset(NumericVector<Number>::build(_dummy_comm).release());
+  _soln_predicted->init(0, 0, false, SERIAL);
+}
+
+StepperInfo::StepperInfo(const StepperInfo& si)
+{
+  _step_count = si._step_count;
+  _time = si._time;
+  _dt = si._dt;
+  _nonlin_iters = si._nonlin_iters;
+  _lin_iters = si._lin_iters;
+  _converged = si._converged;
+  _solve_time_secs = si._solve_time_secs;
+
+  _soln_nonlin.reset(si._soln_nonlin->clone().release());
+  _soln_aux.reset(si._soln_aux->clone().release());
+  _soln_predicted.reset(si._soln_predicted->clone().release());
+
+  _snapshot = false;
+  _rewind = false;
+  _rewind_time = -1;
+}
+
+StepperInfo&
+StepperInfo::operator=(const StepperInfo& si)
+{
+  _step_count = si._step_count;
+  _time = si._time;
+  _dt = si._dt;
+  _nonlin_iters = si._nonlin_iters;
+  _lin_iters = si._lin_iters;
+  _converged = si._converged;
+  _solve_time_secs = si._solve_time_secs;
+
+  _soln_nonlin.reset(si._soln_nonlin->clone().release());
+  _soln_aux.reset(si._soln_aux->clone().release());
+  _soln_predicted.reset(si._soln_predicted->clone().release());
+
+  _snapshot = false;
+  _rewind = false;
+  _rewind_time = -1;
+  return *this;
+}
+
+void
+StepperInfo::pushHistory(Real dt, bool converged, Real solve_time)
+{
+    _dt.push_front(dt);
+    _dt.pop_back();
+    _converged.push_front(converged);
+    _converged.pop_back();
+    _solve_time_secs.push_front(solve_time);
+    _solve_time_secs.pop_back();
+}
+
+void
+StepperInfo::update(
+    int step_count,
+    Real time,
+    Real dt,
+    unsigned int nonlin_iters,
+    unsigned int lin_iters,
+    bool converged,
+    Real solve_time_secs,
+    std::vector<Real> soln_nonlin,
+    std::vector<Real> soln_aux,
+    std::vector<Real> soln_predicted
+    )
+{
+  _step_count = step_count;
+  _time = time;
+  _dt.push_front(dt);
+  _dt.pop_back();
+  _nonlin_iters = nonlin_iters;
+  _lin_iters = lin_iters;
+  _converged.push_front(converged);
+  _converged.pop_back();
+  _solve_time_secs.push_front(solve_time_secs);
+  _solve_time_secs.pop_back();
```

Comment:
I see several push/pops happening in pairs here but I don't understand why. Are we trying to keep this std::lists to a fixed size? The documentation isn't very clear here or in the header.